### PR TITLE
Adding query stack fault to MSQ to capture native query errors.

### PR DESCRIPTION
**Pull request by cryptoe**
#### Description

* Add a new fault "QueryRuntimeError" to MSQ engine to capture native query errors. 
* Fixed bug in MSQ fault tolerance where worker were being retried if `UnexpectedMultiValueDimensionException` was thrown.
* An exception from the query runtime with `org.apache.druid.query` as the package name is thrown as a QueryRuntimeError

#### Release note
Add a new fault "QueryRuntimeError" to MSQ engine to capture native query errors. 
Fixed bug in MSQ fault tolerance where worker were being retried if `UnexpectedMultiValueDimensionException` was thrown.

<hr>

##### Key changed/added classes in this PR
 * `MSQErrorReport`
 * `QueryRuntimeError`

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.

### CODE REVIEW

**Review by gianm — COMMENTED** [review]
There is a new error code `QueryStackError` that is only used in one case: when the query stack throws `UnexpectedMultiValueDimensionException`. Did you mean to associate this error code with more kinds of problems?

If we keep this design, also, to me `QueryRuntimeError` makes more sense than `QueryStackError`. I think end users are more likely to think of the thing that runs queries as a runtime vs a stack. ("Stack of what?")

**Comment by cryptoe** [comment]
I earlier made changes in BaseLeafFrameProcessor#runIncrementally method but since that would also throw FrameSpecific exceptions I dropped that idea. 


> There is a new error code QueryStackError that is only used in one case: when the query stack throws UnexpectedMultiValueDimensionException. Did you mean to associate this error code with more kinds of problems?

I was thinking we would eventually add more errors and `UnexpectedMultiValueDimensionException` is just the start. 

Another approach was to get the exception and check if the exception class name starts with `org.apache.druid.query` . It might be a hack but it may work for about 10 exceptions. WDYT ?

Screenshot 2023-03-14 at 12 42 18 PM

> If we keep this design, also, to me QueryRuntimeError makes more sense than QueryStackError

Sure I can rename `QueryStackError` to `QueryRuntimeError`.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/QueryStackFault.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+package org.apache.druid.msq.indexing.error;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+
+import java.util.Objects;
+
+/**
+ * Fault to throw when the error cames from the druid native query stack while running in the MSQ engine .
```
Comment (abhishekrb19) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/QueryStackFault.java:29]:
```suggestion
 * Fault to throw when the error comes from the druid native query stack while running in the MSQ engine.
```

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/QueryStackFault.java:29]:
```suggestion
 * Encapsulates the native query stack errors
```

**Review by LakshSingla — COMMENTED** [review]
Thanks for the changes 🚀 . Agreeing with @gianm's comments on 2 things:
1. Renaming the fault class name
2. This seems like a one-off case that we handle currently. Can we identify the cases where we defer to the native query engine for our processing, and wrap the errors there in this new Fault. For example, in the `GroupByPreShuffleProcessor`, we create the frame writer and iterate over it. I remember adding a similar fault a while back, and I ran into the same issue so I wrapped a large portion of the code (which might encapsulate the frame errors as well). Is there a way to change the design somehow (maybe in a separate PR), so that we can identify the frame errors v/s rest of the errors easily?

The second point above might not be possible in this PR, so I am okay with the changes, and you can add a comment in the base class or the MSQErrorReport's code to mention the possibility of a better design.

**Review by LakshSingla — APPROVED** [review]
Thanks for adding the tests, and the changes LGTM. We should look for a way to reduce the if..else conditions in the report generation in subsequent PRs.

**Review by kfaraz — APPROVED** [review]
LGTM

**Pull request merged**